### ISSUE TITLE
`fireEvent.press` does not trigger style callback

### ISSUE THREAD

**Issue by Vaeb** [callstack/react-native-testing-library#1272]
## Context
`Pressable`'s `style` property uses the following type:
```
StyleProp<ViewStyle> | ((state: PressableStateCallbackType) => StyleProp<ViewStyle>)
````
which uses:
```
interface PressableStateCallbackType {
    readonly pressed: boolean;
}
```
The callback style allows for briefly changing an element's style after pressing it.

## Describe the bug

Currently `fireEvent.press` does not trigger the style callback. Hence, any statements within a `state.pressed === true` branch cannot be tested if the callback is not exposed.

Example:
```typescript
/* Component */
function Button() {
  const styleCallback = useCallback(
    (state: PressableStateCallbackType) => ({
      backgroundColor: state.pressed ? "#f00" : "#00f"
    }),
    []
  );

  return <Pressable testID="button" style={styleCallback} />
}

/* Test */
it("handles press", async () => {
  const wrapper = render(<Button />);
  const pressable = wrapper.getByTestId("button");
  fireEvent.press(pressable);
});
```

`Button` has full statement coverage but only partial branch coverage (according to Jest). Branch coverage can be forced to 100% by running `styleCallback({ pressed: true });` within the `Button` component. The style callback is never otherwise called with a `{ pressed: true }` value.

## Expected behavior
When a `Pressable`'s `style` is a callback it should be called with a `{ pressed: true }` state upon `fireEvent.press` being called. The style callback would then be called a second time with `{ pressed: false }` to emulate the end of the press event.

In the prior example, this change would result in Jest showing full branch coverage of `Button` instead of partial branch coverage.

## Versions

```
@testing-library/react-native: 11.4.0 => 11.4.0
react: 17.0.1 => 17.0.1
react-native: 0.64.1 => 0.64.1
react-test-renderer: 17.0.1 => 17.0.1
```

**Comment by mdjastrzebski** [comment]
TLDR:
Use `fireEvent(pressable, "pressIn")` and `fireEvent(pressable, "pressOut")`.

Long version:
`Pressable` changes it's internal `pressed` state on [`onPressIn`](https://github.com/facebook/react-native/blob/4ac4a5c27dc5705a42ed7f607e2333d363c0a6c5/Libraries/Components/Pressable/Pressable.js#L295) and `onPressOut` events. That state is used to pass `pressed` argument to [`style` callback](https://github.com/facebook/react-native/blob/4ac4a5c27dc5705a42ed7f607e2333d363c0a6c5/Libraries/Components/Pressable/Pressable.js#L342). Regular `onPress` event does not affect that internal state.

`fireEvent` offers low-level event triggering capability, meaning that it will only fire the exact event you asking it for. We have plans for higher level `userEvent` tool that would perform a series of events that correspond to real React Native interaction but that is still in planning (see #1119). Therefore, you need to call `onPressIn` and `onPressOut` in order to change the internal `Pressable` state instead of `fireEvent.press()`. You can do that by using following syntax: `fireEvent(pressable, "pressIn")` and `fireEvent(pressable, "pressOut")`.

**Comment by Vaeb** [comment]
Thanks for the quick reply. I tried `fireEvent(pressable, "pressIn")` and received the same result as before - no increase in branch coverage or expected style changes. Adding `fireEvent(pressable, "pressOut")` was the same.

Example: 
```typescript
it("handles press", async () => {
  const wrapper = render(<Button />);
  const pressable = wrapper.getByTestId("button");
  fireEvent(pressable, "pressIn");
  expect(pressable).backgroundColor({ borderColor: "#f00" });
  fireEvent(pressable, "pressOut");
  expect(pressable).backgroundColor({ borderColor: "#00f" });
});
```

**Comment by mdjastrzebski** [comment]
Weird it should work. I will have to look deeper into that.

**Comment by mdjastrzebski** [comment]
So far figured out how to trigger pressed state:
```ts
test('handles press', async () => {
  const wrapper = render(<Button />);
  const pressable = wrapper.getByTestId('button');
  fireEvent(pressable, 'onResponderGrant', {
    persist: jest.fn(),
    nativeEvent: {
      timestamp: Date.now(),
    },
  });
  expect(pressable).toHaveStyle({ backgroundColor: '#f00' });
```

**Comment by mdjastrzebski** [comment]
@Vaeb you also have [`testOnly_pressed` prop](https://reactnative.dev/docs/pressable#testonly_pressed) that should trigger the relevant code path.

I think your case is kind of edge, as you target 100% code coverage. For typical tests when you want to test key features of the app, using `fireEvent.press` would be enough as relevant events will be fired. Therefore, I am closing this issue, as proper fix that would satisfy 100% replication of `Pressable` event processing would be quite complex, and imo not worth the added complexity.

**Comment by zigcccc** [comment]
fwiw, using the above snippet from @mdjastrzebski, I augmented the "original" `fireEvent` like so:
```typescript
const fireEvent: typeof rtlFireEvent & { pressWithPersistedState: (Element: ReactTestInstance) => void } = rtlFireEvent;
fireEvent.pressWithPersistedState = (element: ReactTestInstance) => {
  rtlFireEvent.press(element);
  rtlFireEvent(element, 'onResponderGrant', {
    persist: jest.fn(),
    nativeEvent: {
      timestamp: Date.now(),
    },
  });
};
```

this allows me to:
a) use the OG `fireEvent.press` in 99% of the cases
b) where needed, I can call `fireEvent.pressWithPersistedState` and I can make assertions on style/children for when the Pressable is in the pressed state

Not the cleanest solution out there, but it gets the job done where needed.

**Comment by reyalpsirc** [comment]
Extending on @mdjastrzebski , it seems that we can get the non-pressed state back using `onResponderTerminate` combined with `waitFor` on the respective `expect`.
Example:
```
fireEvent(button, 'onResponderTerminate', {
  persist: jest.fn(),
  nativeEvent: {
    timestamp: Date.now(),
  },
});
await waitFor(() => {
  expect(mergeStyles(textLabel.props.style).color).toBe('#ffffff')
})
```
